File: src/main/java/org/voegtle/weatherwidget/util/StatisticsUpdater.java
package org.voegtle.weatherwidget.util;

import android.app.Activity;
import org.voegtle.weatherwidget.data.Statistics;
import org.voegtle.weatherwidget.location.LocationView;
import org.voegtle.weatherwidget.location.WeatherLocation;
import org.voegtle.weatherwidget.state.State;
import org.voegtle.weatherwidget.state.StateCache;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class StatisticsUpdater {
  private StateCache stateCache;

  private WeatherDataFetcher weatherDataFetcher;

  public StatisticsUpdater(Activity activity) {
    this.stateCache = new StateCache(activity);

    this.weatherDataFetcher = new WeatherDataFetcher(ContextUtil.getBuildNumber(activity));

  }

  public void setupStatistics(final LocationView locationView) {
    State state = stateCache.read(locationView.getId());
    locationView.setExpanded(state.isExpanded());
  }

  public void updateStatistics(LocationView locationView, WeatherLocation location) {
    HashMap<LocationView, WeatherLocation> updateCandidates = new HashMap<>();
    updateCandidates.put(locationView, location);
    updateStatistics(updateCandidates, true);
  }

  public void updateStatistics(final HashMap<LocationView, WeatherLocation> updateCandidates, final boolean forceUpdate) {
    updateCachedLocations(updateCandidates);

    final Runnable updater = new Runnable() {
      @Override
      public void run() {
        ArrayList<String> outdatedLocations = lookupOutdatedLocations(updateCandidates, forceUpdate);

        HashMap<String, Statistics> statistics = weatherDataFetcher.fetchStatisticsFromUrl(outdatedLocations);
        updateLocations(updateCandidates, statistics);
      }
    };
    ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
    scheduler.schedule(updater, 0, TimeUnit.SECONDS);
  }


  private void updateLocations(HashMap<LocationView, WeatherLocation> updateCandidates, HashMap<String, Statistics> statistics) {
    for (LocationView locationView : updateCandidates.keySet()) {
      WeatherLocation location = updateCandidates.get(locationView);
      Statistics stats = statistics.get(location.getIdentifier());
      if (stats != null) {
        updateLocation(locationView, stats);
      }

    }
  }

  private ArrayList<String> lookupOutdatedLocations(HashMap<LocationView, WeatherLocation> updateCandidates, boolean forceUpdate) {
    ArrayList<String> outdatedLocations = new ArrayList<>();
    for (LocationView locationView : updateCandidates.keySet()) {
      State state = stateCache.read(locationView.getId());
      if (state.outdated() || forceUpdate) {
        WeatherLocation outdatedLocation = updateCandidates.get(locationView);
        outdatedLocations.add(outdatedLocation.getIdentifier());
      }

    }
    return outdatedLocations;
  }

  private void updateCachedLocations(HashMap<LocationView, WeatherLocation> updateCandidates) {
    for (LocationView locationView : updateCandidates.keySet()) {
      State state = stateCache.read(locationView.getId());
      if (!state.outdated()) {
        updateView(locationView, JsonTranslator.toSingleStatistics(state.getStatistics()));
      }
    }
  }

  private void updateLocation(final LocationView locationView, final Statistics statistics) {
    locationView.post(new Runnable() {
      @Override
      public void run() {
        updateView(locationView, statistics);
        updateCache(locationView, statistics);
      }
    });
  }

  private void updateCache(LocationView locationView, Statistics statistics) {
    State state = new State(locationView.getId());
    state.setAge(new Date());
    state.setExpanded(locationView.isExpanded());
    state.setStatistics(JsonTranslator.toString(statistics));
    stateCache.save(state);
  }

  private void updateView(final LocationView locationView, final Statistics moreData) {
    locationView.setMoreData(moreData);
  }

  public void clearState(LocationView locationView) {
    State state = stateCache.read(locationView.getId());
    state.setExpanded(false);
    state.setStatistics("");
    state.setAge(DateUtil.getYesterday());
    stateCache.save(state);
  }

}

File: src/main/java/org/voegtle/weatherwidget/util/StatisticsUpdater.kt
package org.voegtle.weatherwidget.util

import android.app.Activity
import org.voegtle.weatherwidget.data.Statistics
import org.voegtle.weatherwidget.location.LocationView
import org.voegtle.weatherwidget.location.WeatherLocation
import org.voegtle.weatherwidget.state.State
import org.voegtle.weatherwidget.state.StateCache

import java.util.ArrayList
import java.util.Date
import java.util.HashMap
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

class StatisticsUpdater(activity: Activity) {
  private val stateCache: StateCache

  private val weatherDataFetcher: WeatherDataFetcher

  init {
    this.stateCache = StateCache(activity)

    this.weatherDataFetcher = WeatherDataFetcher(ContextUtil.getBuildNumber(activity))

  }

  fun setupStatistics(locationView: LocationView) {
    val state = stateCache.read(locationView.id)
    locationView.isExpanded = state.isExpanded
  }

  fun updateStatistics(locationView: LocationView, location: WeatherLocation) {
    val updateCandidates = HashMap<LocationView, WeatherLocation>()
    updateCandidates.put(locationView, location)
    updateStatistics(updateCandidates, true)
  }

  fun updateStatistics(updateCandidates: HashMap<LocationView, WeatherLocation>, forceUpdate: Boolean) {
    updateCachedLocations(updateCandidates)

    val updater = Runnable {
      val outdatedLocations = lookupOutdatedLocations(updateCandidates, forceUpdate)

      val statistics = weatherDataFetcher.fetchStatisticsFromUrl(outdatedLocations)
      updateLocations(updateCandidates, statistics)
    }
    val scheduler = Executors.newScheduledThreadPool(1)
    scheduler.schedule(updater, 0, TimeUnit.SECONDS)
  }


  private fun updateLocations(updateCandidates: HashMap<LocationView, WeatherLocation>, statistics: HashMap<String, Statistics>) {
    for (locationView in updateCandidates.keys) {
      val location: WeatherLocation? = updateCandidates[locationView]
      val stats = statistics[location?.identifier]
      if (stats != null) {
        updateLocation(locationView, stats)
      }

    }
  }

  private fun lookupOutdatedLocations(updateCandidates: HashMap<LocationView, WeatherLocation>, forceUpdate: Boolean): ArrayList<String> {
    val outdatedLocations = ArrayList<String>()
    for (locationView in updateCandidates.keys) {
      val state = stateCache.read(locationView.id)
      if (state.outdated() || forceUpdate) {
        val outdatedLocation: WeatherLocation? = updateCandidates[locationView]
        outdatedLocations.add(outdatedLocation!!.identifier!!)
      }

    }
    return outdatedLocations
  }

  private fun updateCachedLocations(updateCandidates: HashMap<LocationView, WeatherLocation>) {
    for (locationView in updateCandidates.keys) {
      val state = stateCache.read(locationView.id)
      if (!state.outdated()) {
        updateView(locationView, JsonTranslator.toSingleStatistics(state.statistics))
      }
    }
  }

  private fun updateLocation(locationView: LocationView, statistics: Statistics) {
    locationView.post {
      updateView(locationView, statistics)
      updateCache(locationView, statistics)
    }
  }

  private fun updateCache(locationView: LocationView, statistics: Statistics) {
    val state = State(locationView.id)
    state.age = Date()
    state.isExpanded = locationView.isExpanded
    state.statistics = JsonTranslator.toString(statistics)
    stateCache.save(state)
  }

  private fun updateView(locationView: LocationView, moreData: Statistics) {
    locationView.setMoreData(moreData)
  }

  fun clearState(locationView: LocationView) {
    val state = stateCache.read(locationView.id)
    state.isExpanded = false
    state.statistics = ""
    state.age = DateUtil.yesterday
    stateCache.save(state)
  }

}
